cdap-app-fabric/src/main/java/co/cask/cdap/route/store/ZKRouteStore.java

```diff
@@ -142,25 +137,66 @@ public void onSuccess(NodeData result) {
         try {
           RouteConfig route = ROUTE_CONFIG_CODEC.decode(result.getData());
           settableFuture.set(route);
-          routeConfigMap.put(serviceId, settableFuture);
         } catch (Exception ex) {
           LOG.debug("Unable to deserialize the config for service {}. Got data {}", serviceId, result.getData());
           // Need to remove the future from the map since later calls will continue to use this future and will think
           // that there is an exception
-          routeConfigMap.remove(serviceId);
+          routeConfigMap.remove(serviceId, settableFuture);
           settableFuture.setException(ex);
         }
       }
 
       @Override
       public void onFailure(Throwable t) {
-        routeConfigMap.remove(serviceId);
         settableFuture.setException(t);
```

Comment:
So, if there is config node, we set the `NoNodeException` into the future. This means every time when the `getConfig` is called (hence calling `future.get`), a new stacktrace is generated, which is quite slow comparing to the no exception case. Maybe better if the node doesn't exists, we set an empty `RouteConfig` into the settable future.